metaflow/plugins/argo/argo_workflows.py

```diff
@@ -0,0 +1,1304 @@
+                }
+            )
+            metaflow_version = self.environment.get_environment_info()
+            metaflow_version["flow_name"] = self.graph.name
+            metaflow_version["production_token"] = self.production_token
+            env["METAFLOW_VERSION"] = json.dumps(metaflow_version)
+
+            # Set the template inputs and outputs for passing state. Very simply,
+            # the container template takes in input-paths as input and outputs
+            # the task-id (which feeds in as input-paths to the subsequent task).
+            # In addition to that, if the parent of the node under consideration
+            # is a for-each node, then we take the split-index as an additional
+            # input. Analogously, if the node under consideration is a foreach
+            # node, then we emit split cardinality as an extra output. I would like
+            # to thank the designers of Argo Workflows for making this so
+            # straightforward!
+            inputs = []
+            if node.name != "start":
+                inputs.append(Parameter("input-paths"))
+            if any(self.graph[n].type == "foreach" for n in node.in_funcs):
+                # Fetch split-index from parent
+                inputs.append(Parameter("split-index"))
+
+            outputs = []
+            if node.name != "end":
+                outputs = [Parameter("task-id").valueFrom({"path": "/mnt/out/task_id"})]
+            if node.type == "foreach":
+                # Emit split cardinality from foreach task
+                outputs.append(
+                    Parameter("num-splits").valueFrom({"path": "/mnt/out/splits"})
+                )
+
+            # Create a ContainerTemplate for this node. Ideally, we would have
+            # liked to inline this ContainerTemplate and avoid scanning the workflow
+            # twice, but due to issues with variable substitution, we will have to
+            # live with this routine.
+            yield (
+                Template(node.name)
+                # Set @timeout values
+                .active_deadline_seconds(run_time_limit)
+                # Set service account
+                .service_account_name(resources["service_account"])
+                # Configure template input
+                .inputs(Inputs().parameters(inputs))
+                # Configure template output
+                .outputs(Outputs().parameters(outputs))
+                # Fail fast!
+                .fail_fast()
+                # Set @retry/@catch values
+                .retry_strategy(
+                    times=total_retries,
+                    minutes_between_retries=minutes_between_retries,
+                )
+                # .metadata(...)
+                # Set emptyDir volume for state management
+                .empty_dir_volume("out")
+                # Set container
+                .container(
+                    # TODO: Unify the logic with kubernetes.py
```

Comment:
Do you want to do this before merging or after?

Reply:
It's not entirely straightforward to unify since not all the fields are available to Argo Workflows. Hopefully, we will learn a bit over the next month or so to identify the right abstraction.